rfcs/RFC: Handling XLA Exceptions in Python

```diff
@@ -0,0 +1,101 @@
+
+1. Though XLA was developed as a part of Tensorflow, it is also used by PyTorch, JAX, etc. There is a need for exceptions thrown by XLA to be named independent of Tensorflow. 
+2. Today, XLA defines XlaRuntimeError in the xla client. However, this exception is not readily available to front-ends. Refer to this (https://github.com/google/jax/pull/10676) effort in JAX to expose this Exception to users.
+
+User Benefit
+
+1. *Simplified contract with front-ends*: Add a single line of code to front-ends to import XLA exceptions. For eg: from import xla.exceptions as xla_exceptions.
+2. *Better exception names in XLA*: Derive readable exceptions like XlaInvalidArgumentError from XlaError to replace the existing ValueError. Additionally, this would make it easier to catch/trap Exceptions from XLA and attribute errors to XLA.
+
+Design Proposal
+
+This proposal would replace existing std:: and py:: exceptions thrown by the CPP layers with new custom exceptions which will all be extended from XlaError base class. All status objects originating from XLA will be replaced by XLA specific status codes. Equivalent python exceptions will be defined under a common module xla.exceptions which will contain translations for status objects (previously TensorFlow exceptions (https://github.com/tensorflow/tensorflow/blob/87462bfac761435a46641ff2f10ad0b6e5414a4b/tensorflow/python/framework/errors_impl.py)) as well as direct throws from XLA (previously std library exceptions).
+
+This proposal introduces an XlaError base class and all exceptions are extended from this class.
+
+class XlaError : public std::Exception {
+public:
+    char * what () {
+        return "Custom XLA Exception";
+    }
+}
+
+We will create different exception classes based on the need by extending this base class. For example:
+
+class XlaInvalidValueError : public XlaError {
+public:
+    char * what () {
+        return "Invalid value provided.";
+    }
+}
+
+We will expose the exceptions under a common module xla.exceptions for use by front-ends. We will additionally modify TensorFlow and PyTorch front-ends to import xla exceptions as import xla.exceptions as xla_exceptions
+
+The effort can be categorized into:
+
+1. Replacing existing status objects
+
+ XlaInternalError (Currently a TensorFlow status object)
+
+Current implementation[Link to code (https://github.com/tensorflow/tensorflow/blob/d1a41c236defc6dd05e6fb31dddea8e0a8b53b96/tensorflow/compiler/xla/runtime/execution_engine.cc#:~:text=return%20InternalError(%22exported%20function%20not%20found%3A%20%25s%22%2C%20function_name)%3B)]:
+
+return InternalError("exported function not found: %s", function_name);
+
+Proposed implementation:
+
+return XlaInternalError("exported function not found: %s", function_name);
+
+Python UX:
+
+import xla.exceptions as xla_exceptions
+try:
+    #Some line of code
+except xla_exceptions.XlaInternalError:
+    pass
+
+2. Replacing existing use of std:: and py:: exceptions in the CPP layer
```

Comment:
Using standard exceptions has the advantage that they can be treated uniformly. I understand that goal of making errors attributable to XLA but in the specific instance called out below, I am not convinced this is the right approach. 
The `__getattr__` method is a standard python method defined on an object that, while returned from an XLA API, is potentially used in contexts independent of the XLA api. The contract for that method is to throw the standard exception so that the object behaves like one would expect of a python object.

So I suggest we only replace exceptions that are an exception of the XLA api in a narrow sense and keep standard exceptions when they model python behavior.

Reply:
What is your take on this @vinayburugu? It seems wrong to me to replace these exceptions from a python API perspective.